timemachine/exchange/exchange_mover.py

```diff
@@ -0,0 +1,354 @@
+# Prototype ExchangeMover that implements an instantaneous water swap move.
+# disable for ~2x speed-up
+from jax import config
+
+config.update("jax_enable_x64", True)
+
+import argparse
+import os
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+from openmm import app
+from rdkit import Chem
+from scipy.stats import special_ortho_group
+
+from timemachine.constants import AVOGADRO, DEFAULT_KT, DEFAULT_PRESSURE, DEFAULT_TEMP
+from timemachine.fe import cif_writer
+from timemachine.fe.free_energy import AbsoluteFreeEnergy, HostConfig, InitialState, image_frames
+from timemachine.fe.topology import BaseTopology
+from timemachine.fe.utils import get_romol_conf
+from timemachine.ff import Forcefield
+from timemachine.ff.handlers.nonbonded import SimpleChargeHandler, SimpleChargeIntraHandler, SimpleChargeSolventHandler
+from timemachine.lib import LangevinIntegrator, MonteCarloBarostat, custom_ops
+from timemachine.md.barostat.utils import get_bond_list, get_group_indices
+from timemachine.md.builders import strip_units
+from timemachine.potentials import nonbonded
+
+# from timemachine.md.minimizer import minimize_host_4d
+
+
+def build_system(host_pdbfile: str, water_ff: str, padding: float):
+    if isinstance(host_pdbfile, str):
+        assert os.path.exists(host_pdbfile)
+        host_pdb = app.PDBFile(host_pdbfile)
+    elif isinstance(host_pdbfile, app.PDBFile):
+        host_pdb = host_pdbfile
+    else:
+        raise TypeError("host_pdbfile must be a string or an openmm PDBFile object")
+
+    host_coords = strip_units(host_pdb.positions)
+
+    box_lengths = np.amax(host_coords, axis=0) - np.amin(host_coords, axis=0)
+
+    box_lengths = box_lengths + padding
+    box = np.eye(3, dtype=np.float64) * box_lengths
+
+    host_ff = app.ForceField(f"{water_ff}.xml")
+    nwa = len(host_coords)
+    solvated_host_system = host_ff.createSystem(
+        host_pdb.topology, nonbondedMethod=app.NoCutoff, constraints=None, rigidWater=False
+    )
+    solvated_host_coords = host_coords
+    solvated_topology = host_pdb.topology
+
+    return solvated_host_system, solvated_host_coords, box, solvated_topology, nwa
+
+
+# currently un-used.
+def randomly_rotate_and_translate(coords, offset):
+    centroid = np.mean(coords, axis=0, keepdims=True)
+    centered_coords = coords - centroid
+    rot_mat = special_ortho_group.rvs(3)
+    rotated_coords = np.matmul(centered_coords, rot_mat)
+    return rotated_coords + offset
+
+
+def randomly_translate(coords, offset):
+    centroid = np.mean(coords, axis=0, keepdims=True)
+    centered_coords = coords - centroid
+    return centered_coords + offset
```

Comment:
Translating to a uniform point in the box lends to a robust implementation, but may be a contributing factor if a lower acceptance rate is observed with this approach compared with prior approaches.

Prior approaches for targeting water displacement moves include:
* Decomposing box into small (0.2A) grid cells, and focusing proposals on sterically unoccupied cells (as in 3.1.1. of https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6456390/ , and GPU implementation in https://pubmed.ncbi.nlm.nih.gov/34762421/ )
* Restricting attention to a spherical region containing both the site to be hydrated and some bulk water (e.g. https://pubmed.ncbi.nlm.nih.gov/32968887/, https://pubs.acs.org/doi/10.1021/acs.jctc.2c00823 , and https://github.com/MobleyLab/blues-water-hopping-paper/blob/master/SI/water-hopping-supplementary-documents/scripts/WaterTranslation.py )

Reply:
Great points.

I was under the impression that that grid approach of 1) was mainly to address the performance issue when PME is used. I'm not sure if it actually modifies the proposal probability in the Metropolis-Hastings evaluation. A spherical selection region could probably be implemented as a Gaussian sphere (whose reverse probability can be computed exactly).

Reply:
>I was under the impression that that grid approach of 1) was mainly to address the performance issue when PME is used. I'm not sure if it actually modifies the proposal probability in the Metropolis-Hastings evaluation.

Hmm, I'm not sure.

(My read of 3.1.1-3.1.2 of https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6456390/ was that the steric grid was different from the grid used for coarse energy evaluations, but I'm not 100% sure, and I didn't catch what the typical sparsities of these two grids were.)